catalog/cataloger_merge.go

```diff
@@ -52,49 +54,46 @@ func (c *cataloger) Merge(ctx context.Context, repository, leftBranch, rightBran
 				Limit: -1,
 			},
 		}
-
-		scanner, err := NewDiffScanner(tx, params)
+		relation, err := getRefsRelationType(tx, params)
 		if err != nil {
 			return nil, err
 		}
-		if scanner.Relation == RelationTypeSame {
+		if relation == RelationTypeSame {
 			return nil, ErrSameBranchMergeNotSupported
 		}
 		nextCommitID, err := getNextCommitID(tx)
 		if err != nil {
 			return nil, err
 		}
-		// do the merge based on the relation
 		previousMaxCommitID, err := getLastCommitIDByBranchID(tx, rightID)
 		if err != nil {
 			return nil, err
 		}
-
-		differences := make(mergeBatchRecords, 0, MergeBatchSize)
+		mergeBatchChan := make(chan mergeBatchRecords, MergeBatchChanBuffer)
+		var workerExitFlag bool
+		go c.diffWorker(params, mergeBatchChan, &workerExitFlag)
 		var rowsCounter int
-		for scanner.Next() {
-			v := scanner.Value()
-			summary[v.Type]++
-			rowsCounter++
-			if v.Type == DifferenceTypeConflict {
-				return nil, ErrConflictFound
+		for buf := range mergeBatchChan {
+			if buf.err != nil {
+				workerExitFlag = true
```

Comment:
This is a race condition per https://golang.org/ref/mem: there is no happens-before relationship between the write here and the read on [line 149](https://github.com/treeverse/lakeFS/pull/887/files#diff-c9a39b7db07f5cc3f023b9c866919ecf8227e2a8fe1702ffcd941f38b9c08fe1R149).
The easiest way to do this is to pass a channel and close it here, and have the worker shut down when the channel closes.  You might use an atomic variable here, but that is more difficult.

Reply:
This is a race condition.  You are quoting when "a read r of a variable v is _allowed_ to observe a write w to v", and "allowed" is italicised.  The very next paragraph states when it _must_ observe the write: " r is _guaranteed_ to observe w if both of the following hold: (1) w happens before r.[...]".  I dropped (2) because (1) doesn't hold: the write to this memory location does not happen before the read.
The burden of proof to show that the write happens-before is on the author not the reviewer.  This is made difficult by this:
> When multiple goroutines access a shared variable v, they must use synchronization events to establish happens-before conditions that ensure reads observe the desired writes
The only synchronization events that I can see are the channel, but that goes in the wrong direction to establish the desired conditions.

It seems to me that to resolve this, we can either run this code through a test with `-race` (coverage says no test covers this case) or we can find a sequence of happens-before conditions in the memory model that shows that the write is before the read or we can rewrite it to use an explicit synchronization mechanism such as closing a channel.

Reply:
Of course the go routine has to read constantly - and it always reads before w wrote. Once w updates it - it will be read and acted upon.(exit the go routine)
A race condition can happen only when more than one party can do something.(one runner can not race himself). This is definitely not the case here.